### DISCUSSION TITLE
Question: Can I prevent FormFields from re-render?

### DISCUSSION THREAD

**Discussion by johnsonthedev** [ariakit/ariakit#1236] [Questions] [answered]
Hi, I might be wrong but I seem to remember you said that FormFields would not re-render when changed. 

But it seems they still do: 

https://codesandbox.io/s/ariakit-form-subscription-0sw5l8?file=/src/App.js

I expected it to behave like this:

https://codesandbox.io/s/valtio-form-ditbgm?file=/src/App.js

Again, it could be that I misunderstood it ;-) Thanks for your help

**Comment by diegohaz** [comment, marked as answer]
I don't know if I can explain this without entering complicated details on how React works.

But, in short, `useFormState` uses React's `useState` underneath. I don't know how Valtio works, but I'd guess the state is handled primarily outside of React (that's what most state management libraries, such as Redux, do to be performant).

Because of that, every update on the Ariakit form state will trigger a re-render on its host component (the component that calls `useFormState()`, which is the `Example` component in your CodeSandbox). Ariakit doesn't do anything custom here. It's just how React works when you have `useState`, `useReducer`, etc. The host component will re-render on state updates.

When a component re-renders, React will also re-render its content by default. Since you're rendering `Field` *inside* `Example`, React will re-render `Field` when `Example` re-renders.

The only way to escape this is `React.memo`ing the `Field` component and passing only the specific props it needs: https://codesandbox.io/s/ariakit-form-memoized-1cq55x?file=/src/App.js:233-237

Ariakit automatically applies `React.memo` to the `FormField` and the `FormInput` components. By default, if they receive only stable references as props, they will re-render only when their respective values change. If you wrap them with a custom component, though, your component will still be re-rendered.

For example, in the GIF below, changing the value of `Name` doesn't re-render `Name2` and vice-versa:

<img src="https://user-images.githubusercontent.com/3068563/163580513-781b4cb8-896c-4429-a8e4-319aa95a3b84.gif" alt="Using form while re-renders are presented as highlights" width="480" />

(I actually found a minor bug here that makes all touched fields re-render, but that doesn't impact your example)

**Reply by johnsonthedev** [reply]
Nice, thanks  !!!  Also for the detailed explanation. Will test it with my forms and check how it performs compared to Valtio, which uses Javascript proxies. Really want to use Ariakit's form fields but use cases seem to be so diverse.
